common/cuda_hip/solver/batch_tridiagonal_solver_kernels.hpp.inc

```diff
@@ -0,0 +1,878 @@
+            }
+
+            subwarp_grp.sync();
+            const int target_lane = i / 2;
+            x_first_higher_group =
+                subwarp_grp.shfl(x_first_higher_group_temp, target_lane);
+        }
+    } else {
+        const int grp_idx = (2 * lane) / final_group_size;
+
+        for (int i = num_grps - 1; i >= 0; i--) {
+            if (grp_idx == i) {
+                my_x_1 = my_d_1 - x_first_higher_group * my_c_1;
+                my_x_2 = my_d_2 - x_first_higher_group * my_c_2;
+            }
+
+            subwarp_grp.sync();
+            const int target_lane = (i * final_group_size) / 2;
+            x_first_higher_group = subwarp_grp.shfl(my_x_1, target_lane);
+        }
+    }
+
+    // Vector writes to enable coalseced accessed while writing
+    if (nrows % 2 == 0) {
+        if (my_row_idx_1 + 1 < nrows) {
+            const ValueType2 my_x_1_2 = make_value_type_2<>(my_x_1, my_x_2);
+            reinterpret_cast<ValueType2*>(x_batch_entry)[my_row_idx_1 / 2] =
+                my_x_1_2;
+        }
+    } else {
+        if (my_row_idx_1 < nrows) {
+            x_batch_entry[my_row_idx_1] = my_x_1;
+        }
+        if (my_row_idx_1 + 1 < nrows) {
+            x_batch_entry[my_row_idx_1 + 1] = my_x_2;
+        }
+    }
+}
+
+
+}  // namespace WM_pGE_approach2
+
+
+template <typename ValueType, int subwarp_size>
+__global__ __launch_bounds__(default_block_size) void WM_pGE_kernel_approach_2(
+    const int num_WM_steps, const size_type nbatch, const int nrows,
+    const ValueType* const __restrict__ a,
+    const ValueType* const __restrict__ b, ValueType* const __restrict__ c,
+    ValueType* const __restrict__ d, ValueType* const __restrict__ x)
+{
+    using Valuetype2 = WM_pGE_approach2::ValueType2<ValueType>;
+
+    auto subwarpgrp =
+        group::tiled_partition<subwarp_size>(group::this_thread_block());
+    const int subgrpwarp_id_in_grid =
+        thread::get_subwarp_id_flat<subwarp_size, int>();
+    const int total_num_subwarp_grps_in_grid =
+        thread::get_subwarp_num_flat<subwarp_size, int>();
+    const int id_within_warp = subwarpgrp.thread_rank();
+
+    // a subwarp per matrix in the batch
+    for (size_type batch_idx = subgrpwarp_id_in_grid; batch_idx < nbatch;
+         batch_idx += total_num_subwarp_grps_in_grid) {
+        // Approach: a thread in the subwarp handles two rows of the matrix or
+        // to be precise, of the matrix tile
+        const auto tile_size = 2 * subwarp_size;
+        const auto num_tiles = ceildiv(nrows, tile_size);
+        const bool is_last_tile_similar = ((nrows % tile_size) == 0);
+        const int final_group_size = pow(2, num_WM_steps);
+        if (final_group_size > tile_size) {
+            printf(
+                "\n Number of WM steps ans subwarp size should satisfy the "
+                "inequality: pow(2, WM_steps) <= tile_size, where tile_size = "
+                "2 "
+                "* subwarp_size");
```

Comment:
Same here as well. Can be moved to the kernel call location, or maybe even to the core function.